File: projects/iffywow/go/romanization.go
package ithkuil

// Romanization profile core-v1. Normative: ../ROMANIZATION.md; the
// conformance vectors in ../conformance/latin_to_coordinate.jsonl are the
// arbiter.
//
// core-v1 accepts exactly one word shape — an unconcatenated formative
// `Vv Cr Vr Ca Vc` (slots II, III, IV, VI, IX) — with closed tables for
// each slot. Everything else (Slot I Cc, affixes, VnCn, stress, glottal
// stop, adjuncts, multiple words, ...) is rejected with ErrUnsupported; the
// profile never guesses.
//
// Every non-ASCII literal below is a single PRECOMPOSED codepoint (NFC),
// exactly as inventoried in ROMANIZATION.md §2 — never a base letter plus a
// combining mark (which would not even compile as a rune literal).

import "strings"

const maxRootCluster = 4

// ---------------------------------------------------------------------------
// Preprocessing (ROMANIZATION.md §1): trim -> NFC -> lowercase.
// ---------------------------------------------------------------------------

// composeTable maps the decomposed (base rune, combining mark) sequences of
// the profile alphabet to their precomposed NFC codepoints.
//
// The Go standard library has no NFC normalizer (that lives in
// golang.org/x/text, and this SDK is stdlib-only), so full NFC is
// approximated by targeted composition over exactly the ROMANIZATION.md §2
// inventory — every profile character plus the stress-marked vowels of
// §2.2, in both cases (composition runs before lowercasing per §1). Any
// sequence outside this table is left untouched and is then rejected as
// ErrUnsupported by classification, which is the required outcome for all
// out-of-alphabet input anyway, so the approximation is behaviorally exact
// for the profile.
var composeTable = map[[2]rune]rune{
	// U+0308 combining diaeresis
	{'a', 0x0308}: 'ä', // ä
	{'e', 0x0308}: 'ë', // ë
	{'o', 0x0308}: 'ö', // ö
	{'u', 0x0308}: 'ü', // ü
	{'A', 0x0308}: 'Ä', // Ä
	{'E', 0x0308}: 'Ë', // Ë
	{'O', 0x0308}: 'Ö', // Ö
	{'U', 0x0308}: 'Ü', // Ü
	// U+0301 combining acute (stress marking; composed so it classifies as
	// a stress vowel rather than a stray combining mark)
	{'a', 0x0301}: 'á', // á
	{'e', 0x0301}: 'é', // é
	{'i', 0x0301}: 'í', // í
	{'o', 0x0301}: 'ó', // ó
	{'u', 0x0301}: 'ú', // ú
	{'A', 0x0301}: 'Á', // Á
	{'E', 0x0301}: 'É', // É
	{'I', 0x0301}: 'Í', // Í
	{'O', 0x0301}: 'Ó', // Ó
	{'U', 0x0301}: 'Ú', // Ú
	// U+0302 combining circumflex (stress marking)
	{'a', 0x0302}: 'â', // â
	{'e', 0x0302}: 'ê', // ê
	{'i', 0x0302}: 'î', // î
	{'o', 0x0302}: 'ô', // ô
	{'u', 0x0302}: 'û', // û
	{'A', 0x0302}: 'Â', // Â
	{'E', 0x0302}: 'Ê', // Ê
	{'I', 0x0302}: 'Î', // Î
	{'O', 0x0302}: 'Ô', // Ô
	{'U', 0x0302}: 'Û', // Û
	// U+030C combining caron
	{'s', 0x030C}: 'š', // š
	{'z', 0x030C}: 'ž', // ž
	{'c', 0x030C}: 'č', // č
	{'r', 0x030C}: 'ř', // ř
	{'n', 0x030C}: 'ň', // ň
	{'S', 0x030C}: 'Š', // Š
	{'Z', 0x030C}: 'Ž', // Ž
	{'C', 0x030C}: 'Č', // Č
	{'R', 0x030C}: 'Ř', // Ř
	{'N', 0x030C}: 'Ň', // Ň
	// U+0327 combining cedilla
	{'t', 0x0327}: 'ţ', // ţ
	{'d', 0x0327}: 'ḑ', // ḑ
	{'c', 0x0327}: 'ç', // ç
	{'l', 0x0327}: 'ļ', // ļ
	{'T', 0x0327}: 'Ţ', // Ţ
	{'D', 0x0327}: 'Ḑ', // Ḑ
	{'C', 0x0327}: 'Ç', // Ç
	{'L', 0x0327}: 'Ļ', // Ļ
	// U+0323 combining dot below
	{'z', 0x0323}: 'ẓ', // ẓ
	{'Z', 0x0323}: 'Ẓ', // Ẓ
}

// composeProfile performs the targeted NFC composition described at
// composeTable. Greedy left-to-right: each rune absorbs following combining
// marks as long as a composition exists.
func composeProfile(s string) string {
	runes := []rune(s)
	out := make([]rune, 0, len(runes))
	for i := 0; i < len(runes); {
		r := runes[i]
		i++
		for i < len(runes) {
			c, ok := composeTable[[2]rune{r, runes[i]}]
			if !ok {
				break
			}
			r = c
			i++
		}
		out = append(out, r)
	}
	return string(out)
}

// ---------------------------------------------------------------------------
// Character inventories (ROMANIZATION.md §2; exact codepoints, NFC).
// ---------------------------------------------------------------------------

// vowelSet — the 9 vowel codepoints of §2.1.
var vowelSet = map[rune]bool{
	'a': true,
	'e': true,
	'i': true,
	'o': true,
	'u': true,
	'ä': true, // ä U+00E4
	'ë': true, // ë U+00EB
	'ö': true, // ö U+00F6
	'ü': true, // ü U+00FC
}

// stressVowelSet — §2.2, always rejected (Slot X stress).
var stressVowelSet = map[rune]bool{
	'á': true, // á
	'é': true, // é
	'í': true, // í
	'ó': true, // ó
	'ú': true, // ú
	'â': true, // â
	'ê': true, // ê
	'î': true, // î
	'ô': true, // ô
	'û': true, // û
}

// rootConsonants — the 27-consonant root inventory in the NORMATIVE order
// of §2.3: the bijective base-27 digit of a consonant is its index here
// plus one.
var rootConsonants = []rune{
	'p', 'b', 't', 'd', 'k', 'g', 'f', 'v', 'ţ', // 1-9:   p b t d k g f v ţ
	'ḑ', 's', 'z', 'c', 'ẓ', 'š', 'ž', 'č', 'j', // 10-18: ḑ s z c ẓ š ž č j
	'ç', 'x', 'ļ', 'l', 'r', 'ř', 'm', 'n', 'ň', // 19-27: ç x ļ l r ř m n ň
}

// rootDigit maps a root consonant to its 1-based digit value.
var rootDigit = func() map[rune]int {
	m := make(map[rune]int, len(rootConsonants))
	for i, r := range rootConsonants {
		m[r] = i + 1
	}
	return m
}()

// segmenterOnly — characters that classify as consonants for run
// segmentation but are valid only where a table row admits them (w/y in Ca)
// or nowhere (h, ').
var segmenterOnly = map[rune]bool{'w': true, 'y': true, 'h': true, '\'': true}

// ---------------------------------------------------------------------------
// Lookup tables (ROMANIZATION.md §4; closed, exact rows).
// ---------------------------------------------------------------------------

type vvEntry struct{ stem, version int } // version: 0 = PRC, 1 = CPT

var vvTable = map[rune]vvEntry{
	'a': {1, 0},
	'ä': {1, 1}, // ä
	'e': {2, 0},
	'i': {2, 1},
	'u': {3, 0},
	'ü': {3, 1}, // ü
	'o': {0, 0},
	'ö': {0, 1}, // ö
}

var vvInverse = func() map[vvEntry]rune {
	m := make(map[vvEntry]rune, len(vvTable))
	for k, v := range vvTable {
		m[v] = k
	}
	return m
}()

type vrEntry struct{ function, spec int } // function: 0 = STA, 1 = DYN

var vrTable = map[rune]vrEntry{
	'a': {0, 0},
	'ä': {0, 1}, // ä
	'e': {0, 2},
	'i': {0, 3},
	'u': {1, 0},
	'ü': {1, 1}, // ü
	'o': {1, 2},
	'ö': {1, 3}, // ö
}

var vrInverse = func() map[vrEntry]rune {
	m := make(map[vrEntry]rune, len(vrTable))
	for k, v := range vrTable {
		m[v] = k
	}
	return m
}()

// caTable: perspective (configuration UNI, extension DEL, affiliation CSL,
// essence NRM).
var caTable = map[rune]int{'l': 0, 'r': 1, 'w': 2, 'y': 3}

var caInverse = func() map[int]rune {
	m := make(map[int]rune, len(caTable))
	for k, v := range caTable {
		m[v] = k
	}
	return m
}()

// vcForms: case index 0..8 (THM INS ABS AFF STM EFF ERG DAT IND). Index 4
// is the two-codepoint digraph "ëi", matched as the whole vowel run.
var vcForms = []string{
	"a",
	"ä", // ä
	"e",
	"i",
	"ëi", // ëi
	"ö",  // ö
	"o",
	"ü", // ü
	"u",
}

// ---------------------------------------------------------------------------
// Segmentation (ROMANIZATION.md §3).
// ---------------------------------------------------------------------------

type latinRun struct {
	vowel bool
	text  []rune
}

func classifyRune(r rune) (vowel bool, err error) {
	switch {
	case stressVowelSet[r]:
		return false, errf(ErrUnsupported,
			"stress-marked vowel %q (Slot X) is outside profile core-v1", r)
	case vowelSet[r]:
		return true, nil
	case rootDigit[r] != 0 || segmenterOnly[r]:
		return false, nil
	default:
		return false, errf(ErrUnsupported, "character %q is outside the profile inventories", r)
	}
}

func segmentLatin(word string) ([]latinRun, error) {
	var runs []latinRun
	for _, r := range word {
		vowel, err := classifyRune(r)
		if err != nil {
			return nil, err
		}
		if n := len(runs); n > 0 && runs[n-1].vowel == vowel {
			runs[n-1].text = append(runs[n-1].text, r)
		} else {
			runs = append(runs, latinRun{vowel: vowel, text: []rune{r}})
		}
	}
	return runs, nil
}

// ---------------------------------------------------------------------------
// FromLatin / ToLatin.
// ---------------------------------------------------------------------------

// FromLatin parses romanized New Ithkuil text (profile core-v1) into a
// canonical Word. Input is preprocessed per ROMANIZATION.md §1: trim,
// Unicode NFC (approximated over the profile alphabet, see composeTable),
// then lowercase. Everything outside the profile rejects with
// ErrUnsupported.
func FromLatin(text string) (Word, error) {
	word := strings.ToLower(composeProfile(strings.TrimSpace(text)))
	if word == "" {
		return Word{}, errf(ErrUnsupported, "empty word")
	}
	runs, err := segmentLatin(word)
	if err != nil {
		return Word{}, err
	}
	// Exactly five runs, alternating, vowel first: Vv Cr Vr Ca Vc.
	if len(runs) != 5 || !runs[0].vowel || runs[1].vowel || !runs[2].vowel ||
		runs[3].vowel || !runs[4].vowel {
		return Word{}, errf(ErrUnsupported,
			"word shape must be Vv Cr Vr Ca Vc (five alternating runs, vowel first), got %d run(s)",
			len(runs))
	}
	vvRun, crRun, vrRun, caRun, vcRun := runs[0].text, runs[1].text, runs[2].text, runs[3].text, runs[4].text

	// Vv — stem + version (§4.1); one character.
	if len(vvRun) != 1 {
		return Word{}, errf(ErrUnsupported, "Vv %q is not a single stem/version vowel", string(vvRun))
	}
	vv, ok := vvTable[vvRun[0]]
	if !ok {
		return Word{}, errf(ErrUnsupported, "Vv %q has no stem/version row", string(vvRun))
	}

	// Cr — root cluster as a bijective base-27 numeral (§5); 1..4 root
	// consonants.
	if len(crRun) > maxRootCluster {
		return Word{}, errf(ErrUnsupported,
			"root cluster %q is longer than %d consonants", string(crRun), maxRootCluster)
	}
	base := uint64(0)
	for _, c := range crRun {
		d, ok := rootDigit[c]
		if !ok {
			return Word{}, errf(ErrUnsupported,
				"consonant %q is not in the root inventory", c)
		}
		base = base*27 + uint64(d)
	}

	// Vr — function + specification (§4.2); one character.
	if len(vrRun) != 1 {
		return Word{}, errf(ErrUnsupported, "Vr %q is not a single function/specification vowel", string(vrRun))
	}
	vr, ok := vrTable[vrRun[0]]
	if !ok {
		return Word{}, errf(ErrUnsupported, "Vr %q has no function/specification row", string(vrRun))
	}

	// Ca — perspective (§4.3); one character.
	if len(caRun) != 1 {
		return Word{}, errf(ErrUnsupported, "Ca %q is not a single perspective consonant", string(caRun))
	}
	perspective, ok := caTable[caRun[0]]
	if !ok {
		return Word{}, errf(ErrUnsupported, "Ca %q is not one of l/r/w/y", string(caRun))
	}

	// Vc — case (§4.4); the whole vowel run must match one form.
	vcStr := string(vcRun)
	caseIndex := -1
	for i, form := range vcForms {
		if form == vcStr {
			caseIndex = i
			break
		}
	}
	if caseIndex < 0 {
		return Word{}, errf(ErrUnsupported, "Vc %q is not a case vowel form of the profile", vcStr)
	}

	// §6 morpheme -> coordinate mapping (provisional, normative for v1).
	return Word{Version: 1, Glyphs: []Glyph{{
		CharacterClass: 0, // formative
		Base:           base,
		Orientation:    vv.stem,
		Sockets: []Socket{
			{ID: 0, Modifier: Modifier{Shape: uint64(vr.spec), Orientation: vr.function + 2*vv.version}},
			{ID: 1, Modifier: Modifier{Shape: uint64(perspective), Orientation: 0}},
			{ID: 2, Modifier: Modifier{Shape: uint64(caseIndex), Orientation: 0}},
		},
	}}}, nil
}

// ToLatin emits the canonical spelling (ROMANIZATION.md §7) of a coordinate
// of exactly the §6 shape. The input is canonicalized first (native input
// is lenient per SDK-INTERFACE.md §2); anything outside the profile shape
// rejects with ErrUnsupported. The result is trimmed, NFC, lowercase.
func ToLatin(w Word) (string, error) {
	c, err := Canonicalize(w)
	if err != nil {
		return "", err
	}
	if c.Version != 1 || len(c.Glyphs) != 1 {
		return "", errf(ErrUnsupported,
			"profile core-v1 covers exactly one glyph at schema version 1, got version %d with %d glyph(s)",
			c.Version, len(c.Glyphs))
	}
	g := c.Glyphs[0]
	if g.CharacterClass != 0 {
		return "", errf(ErrUnsupported,
			"character class %d has no romanization in profile core-v1", g.CharacterClass)
	}
	if len(g.Sockets) != 3 || g.Sockets[0].ID != 0 || g.Sockets[1].ID != 1 || g.Sockets[2].ID != 2 {
		return "", errf(ErrUnsupported, "profile core-v1 requires exactly sockets 0, 1, 2")
	}
	m0 := g.Sockets[0].Modifier
	m1 := g.Sockets[1].Modifier
	m2 := g.Sockets[2].Modifier
	for i, m := range []Modifier{m0, m1, m2} {
		if len(m.Diacritics) != 0 || len(m.Sockets) != 0 {
			return "", errf(ErrUnsupported,
				"socket %d: diacritics and child sockets have no romanization in profile core-v1", i)
		}
	}
	if m0.Shape > 3 { // specification
		return "", errf(ErrUnsupported, "socket 0 shape %d is outside the Vr table", m0.Shape)
	}
	if m1.Shape > 3 || m1.Orientation != 0 { // perspective
		return "", errf(ErrUnsupported, "socket 1 must be a perspective modifier (shape 0..3, orientation 0)")
	}
	if m2.Shape > 8 || m2.Orientation != 0 { // case
		return "", errf(ErrUnsupported, "socket 2 must be a case modifier (shape 0..8, orientation 0)")
	}
	if g.Base < 1 {
		return "", errf(ErrUnsupported,
			"root code 0 has no consonant cluster (bijective numeration starts at 1)")
	}
	cluster, err := decodeRootCluster(g.Base)
	if err != nil {
		return "", err
	}
	fnVer := m0.Orientation // 0..3, range-checked by Canonicalize
	vv := vvInverse[vvEntry{stem: g.Orientation, version: fnVer / 2}]
	vr := vrInverse[vrEntry{function: fnVer % 2, spec: int(m0.Shape)}]
	ca := caInverse[int(m1.Shape)]
	vc := vcForms[int(m2.Shape)]
	return string(vv) + cluster + string(vr) + string(ca) + vc, nil
}

// decodeRootCluster inverts the bijective base-27 numeration of
// ROMANIZATION.md §5: digit = ((n−1) mod 27) + 1, prepend consonant #digit,
// n = (n−1) div 27. Rejects codes decoding to more than four consonants
// (valid root codes are 1..551880).
func decodeRootCluster(base uint64) (string, error) {
	var cluster []rune
	for n := base; n > 0; {
		if len(cluster) == maxRootCluster {
			return "", errf(ErrUnsupported,
				"root code %d decodes to more than %d consonants", base, maxRootCluster)
		}
		d := (n-1)%27 + 1
		cluster = append([]rune{rootConsonants[d-1]}, cluster...)
		n = (n - 1) / 27
	}
	return string(cluster), nil
}

File: projects/iffywow/node/src/romanization.js
/**
 * Romanization profile core-v1 — exact implementation of ../../ROMANIZATION.md.
 *
 * core-v1 accepts exactly one word shape: an unconcatenated formative
 * `Vv Cr Vr Ca Vc` (slots II, III, IV, VI, IX) with closed tables per slot.
 * Everything else — Slot I Cc, affixes, VnCn, stress, glottal stop,
 * adjuncts, multiple words — is rejected with error code `unsupported`;
 * the profile never guesses.
 *
 * Every non-ASCII inventory codepoint below is constructed from its numeric
 * codepoint (String.fromCodePoint) so the tables are guaranteed to hold
 * single precomposed NFC codepoints regardless of how this source file is
 * normalized on disk.
 */

import { IthkuilError } from "./errors.js";
import { validate, canonicalize } from "./coord.js";

function unsupported(detail) {
  return new IthkuilError("unsupported", detail);
}

// --------------------------------------------------------------------------
// §2 Character inventories (exact codepoints, NFC).
// --------------------------------------------------------------------------

// Built from numeric codepoints (not source literals) so the tables hold the
// exact single precomposed NFC codepoints of ROMANIZATION.md SS2 no matter how
// this file is normalized on disk.
const chr = (codepoint) => String.fromCodePoint(codepoint);

const A_UML = chr(0x00e4); // a-diaeresis
const E_UML = chr(0x00eb); // e-diaeresis
const O_UML = chr(0x00f6); // o-diaeresis
const U_UML = chr(0x00fc); // u-diaeresis

/** §2.1 Vowels (class V) — 9 codepoints. */
const VOWEL_SET = new Set(["a", "e", "i", "o", "u", A_UML, E_UML, O_UML, U_UML]);

/**
 * §2.3 Root inventory — 27 consonants in NORMATIVE order; digit value = index + 1.
 * (Defines the bijective base-27 digits of §5.)
 */
const ROOT_CONSONANTS = Object.freeze([
  "p", //           1
  "b", //           2
  "t", //           3
  "d", //           4
  "k", //           5
  "g", //           6
  "f", //           7
  "v", //           8
  chr(0x0163), //   9  t-cedilla
  chr(0x1e11), //  10  d-cedilla
  "s", //          11
  "z", //          12
  "c", //          13
  chr(0x1e93), //  14  z-dot-below
  chr(0x0161), //  15  s-caron
  chr(0x017e), //  16  z-caron
  chr(0x010d), //  17  c-caron
  "j", //          18
  chr(0x00e7), //  19  c-cedilla
  "x", //          20
  chr(0x013c), //  21  l-cedilla
  "l", //          22
  "r", //          23
  chr(0x0159), //  24  r-caron
  "m", //          25
  "n", //          26
  chr(0x0148), //  27  n-caron
]);

const ROOT_DIGIT = new Map(ROOT_CONSONANTS.map((c, i) => [c, i + 1]));

/**
 * Additional segmenter consonants (§2.3): classify as C for run segmentation
 * but are valid only where a table row admits them (`w`/`y`: Ca only;
 * `h`/`'`: nowhere).
 */
const SEGMENT_CONSONANTS = new Set([...ROOT_CONSONANTS, "w", "y", "h", "'"]);

// --------------------------------------------------------------------------
// §4 Lookup tables (closed; exact rows).
// --------------------------------------------------------------------------

/** §4.1 Vv — stem + version (Slot II). value = [stem, version]. */
const VV = new Map([
  ["a", [1, 0]],
  [A_UML, [1, 1]],
  ["e", [2, 0]],
  ["i", [2, 1]],
  ["u", [3, 0]],
  [U_UML, [3, 1]],
  ["o", [0, 0]],
  [O_UML, [0, 1]],
]);

/** Vv inverse: VV_INVERSE[stem][version]. */
const VV_INVERSE = [
  ["o", O_UML], // stem 0
  ["a", A_UML], // stem 1
  ["e", "i"], //   stem 2
  ["u", U_UML], // stem 3
];

/** §4.2 Vr — function + specification (Slot IV). value = [function, spec]. */
const VR = new Map([
  ["a", [0, 0]],
  [A_UML, [0, 1]],
  ["e", [0, 2]],
  ["i", [0, 3]],
  ["u", [1, 0]],
  [U_UML, [1, 1]],
  ["o", [1, 2]],
  [O_UML, [1, 3]],
]);

/** Vr inverse: VR_INVERSE[function][spec]. */
const VR_INVERSE = [
  ["a", A_UML, "e", "i"], // STA
  ["u", U_UML, "o", O_UML], // DYN
];

/** §4.3 Ca — perspective (Slot VI). */
const CA = new Map([
  ["l", 0], // M  monadic
  ["r", 1], // G  agglomerative
  ["w", 2], // N  nomic
  ["y", 3], // A  abstract
]);

const CA_INVERSE = ["l", "r", "w", "y"];

/** §4.4 Vc — case (Slot IX; cases 1-9). `ëi` is U+00EB U+0069, matched as the whole run. */
const VC = new Map([
  ["a", 0], //         THM
  [A_UML, 1], //       INS
  ["e", 2], //         ABS
  ["i", 3], //         AFF
  [E_UML + "i", 4], // STM (digraph ëi)
  [O_UML, 5], //       EFF
  ["o", 6], //         ERG
  [U_UML, 7], //       DAT
  ["u", 8], //         IND
]);

const VC_INVERSE = ["a", A_UML, "e", "i", E_UML + "i", O_UML, "o", U_UML, "u"];

// --------------------------------------------------------------------------
// §5 Root cluster <-> base-27 integer (bijective numeration, no zero digit).
// --------------------------------------------------------------------------

export const MAX_ROOT_CODE = 551880; // 27 + 27^2 + 27^3 + 27^4 (k <= 4)

function encodeRoot(cluster) {
  const chars = [...cluster];
  if (chars.length > 4) {
    throw unsupported(`root cluster "${cluster}" is longer than 4 consonants`);
  }
  let acc = 0;
  for (const ch of chars) {
    const digit = ROOT_DIGIT.get(ch);
    if (digit === undefined) {
      throw unsupported(`consonant "${ch}" (U+${cp(ch)}) is not in the 27-consonant root inventory`);
    }
    acc = acc * 27 + digit;
  }
  return acc; // 1 <= acc <= 551880
}

function decodeRoot(base) {
  if (base < 1 || base > MAX_ROOT_CODE) {
    throw unsupported(`root code ${base} is outside 1..${MAX_ROOT_CODE} (clusters of 1..4 consonants)`);
  }
  let n = base;
  const out = [];
  while (n > 0) {
    const digit = ((n - 1) % 27) + 1; // 1..27
    out.unshift(ROOT_CONSONANTS[digit - 1]);
    n = Math.floor((n - 1) / 27);
  }
  return out.join(""); // 1..4 consonants, guaranteed by the range check
}

function cp(ch) {
  return ch.codePointAt(0).toString(16).toUpperCase().padStart(4, "0");
}

// --------------------------------------------------------------------------
// from_latin (§1 preprocessing, §3 segmentation, §4-§6 tables and mapping).
// --------------------------------------------------------------------------

/**
 * Romanized New Ithkuil text -> canonical, deep-frozen Coord.
 * Rejects everything outside profile core-v1 with code `unsupported`.
 * @param {string} text
 */
export function fromLatin(text) {
  if (typeof text !== "string") {
    throw new IthkuilError("invalid_structure", `romanized input must be a string, got ${typeof text}`);
  }
  // §1: trim -> NFC -> lowercase, in this exact order.
  const word = text.trim().normalize("NFC").toLowerCase();
  if (word.length === 0) throw unsupported("empty word");

  // §3: maximal same-class runs; anything outside the inventories rejects.
  const runs = [];
  for (const ch of word) {
    let cls;
    if (VOWEL_SET.has(ch)) cls = "V";
    else if (SEGMENT_CONSONANTS.has(ch)) cls = "C";
    else throw unsupported(`character "${ch}" (U+${cp(ch)}) is outside the core-v1 inventories`);
    const last = runs[runs.length - 1];
    if (last !== undefined && last.cls === cls) last.text += ch;
    else runs.push({ cls, text: ch });
  }
  // Maximal runs alternate by construction: five runs starting with V is
  // exactly the shape V C V C V.
  if (runs.length !== 5 || runs[0].cls !== "V") {
    throw unsupported(
      `word shape must be Vv Cr Vr Ca Vc (five alternating runs, vowel first); got ${runs.length} run(s)` +
        ` starting with ${runs[0].cls === "V" ? "a vowel" : "a consonant"}`,
    );
  }
  const [vv, cr, vr, ca, vc] = runs.map((r) => r.text);

  // §4.1 Vv — one character, a key of the Vv table.
  const vvRow = VV.get(vv);
  if (vvRow === undefined) throw unsupported(`Vv "${vv}" is not a stem/version vowel of the profile`);
  const [stem, version] = vvRow;

  // §5 Cr — 1..4 root-inventory consonants, bijective base-27.
  const base = encodeRoot(cr);

  // §4.2 Vr — one character, a key of the Vr table.
  const vrRow = VR.get(vr);
  if (vrRow === undefined) throw unsupported(`Vr "${vr}" is not a function/specification vowel of the profile`);
  const [fn, spec] = vrRow;

  // §4.3 Ca — one of l/r/w/y.
  const perspective = CA.get(ca);
  if (perspective === undefined) throw unsupported(`Ca "${ca}" is not a default Ca form (l, r, w, y)`);

  // §4.4 Vc — the whole vowel run must be a Vc form (incl. the digraph "ëi").
  const caseIndex = VC.get(vc);
  if (caseIndex === undefined) throw unsupported(`Vc "${vc}" is not a case vowel form of the profile`);

  // §6: one glyph, three sockets 0/1/2, empty diacritics and child sockets.
  return validate({
    version: 1,
    glyphs: [
      {
        characterClass: 0,
        base,
        orientation: stem,
        sockets: [
          { id: 0, modifier: { shape: spec, orientation: fn + 2 * version, diacritics: [], sockets: [] } },
          { id: 1, modifier: { shape: perspective, orientation: 0, diacritics: [], sockets: [] } },
          { id: 2, modifier: { shape: caseIndex, orientation: 0, diacritics: [], sockets: [] } },
        ],
      },
    ],
  });
}

// --------------------------------------------------------------------------
// to_latin (§7 canonical spelling).
// --------------------------------------------------------------------------

/**
 * Canonical Coord -> canonical spelling (trimmed, NFC, lowercase).
 * Accepts EXACTLY coordinates of the §6/§7 shape; everything else is
 * `unsupported`. Native input is canonicalized first (sockets sorted,
 * empties dropped) per SDK-INTERFACE.md.
 * @returns {string}
 */
export function toLatin(coord) {
  const word = canonicalize(coord);
  if (word.version !== 1) {
    throw unsupported(`romanization profile core-v1 is defined for schema version 1, got ${word.version}`);
  }
  if (word.glyphs.length !== 1) {
    throw unsupported(`core-v1 words have exactly one glyph, got ${word.glyphs.length}`);
  }
  const g = word.glyphs[0];
  if (g.characterClass !== 0) {
    throw unsupported(`core-v1 glyphs have character class 0 (formative), got ${g.characterClass}`);
  }
  if (g.base < 1 || g.base > MAX_ROOT_CODE) {
    throw unsupported(`root code ${g.base} is outside 1..${MAX_ROOT_CODE}`);
  }
  const stem = g.orientation; // 0..3 guaranteed by canonicalize
  if (g.sockets.length !== 3 || g.sockets[0].id !== 0 || g.sockets[1].id !== 1 || g.sockets[2].id !== 2) {
    throw unsupported("core-v1 glyphs carry exactly the three sockets 0, 1, 2");
  }
  const [m0, m1, m2] = g.sockets.map((s) => s.modifier);
  for (const m of [m0, m1, m2]) {
    if (m.diacritics.length !== 0 || m.sockets.length !== 0) {
      throw unsupported("core-v1 modifiers have no diacritics and no child sockets");
    }
  }
  if (m0.shape > 3) throw unsupported(`specification (socket 0 shape) must be 0..3, got ${m0.shape}`);
  if (m1.shape > 3 || m1.orientation !== 0) {
    throw unsupported("perspective modifier (socket 1) must have shape 0..3 and orientation 0");
  }
  if (m2.shape > 8 || m2.orientation !== 0) {
    throw unsupported("case modifier (socket 2) must have shape 0..8 and orientation 0");
  }

  const fnVer = m0.orientation; // 0..3 guaranteed by canonicalize
  const version = Math.floor(fnVer / 2);
  const fn = fnVer % 2;

  const vv = VV_INVERSE[stem][version];
  const cr = decodeRoot(g.base);
  const vr = VR_INVERSE[fn][m0.shape];
  const ca = CA_INVERSE[m1.shape];
  const vc = VC_INVERSE[m2.shape];
  return vv + cr + vr + ca + vc; // all-lowercase, NFC, no separators
}
